src/pool/srv_target.c

```diff
@@ -722,98 +734,62 @@ pool_tgt_query(struct ds_pool *pool, struct daos_pool_space *ps)
 	return rc;
 }
 
-void
-ds_pool_tgt_connect_handler(crt_rpc_t *rpc)
+int
+ds_pool_tgt_connect(struct ds_pool *pool, struct pool_iv_conn *pic)
 {
-	struct pool_tgt_connect_in	*in = crt_req_get(rpc);
-	struct pool_tgt_connect_out	*out = crt_reply_get(rpc);
-	struct ds_pool			*pool;
-	struct ds_pool_hdl		*hdl;
-	int				 rc;
-
-	D_DEBUG(DF_DSMS, DF_UUID": handling rpc %p: hdl="DF_UUID"\n",
-		DP_UUID(in->tci_uuid), rpc, DP_UUID(in->tci_hdl));
+	struct ds_pool_hdl	*hdl = NULL;
+	d_iov_t			cred_iov;
+	int			rc;
 
-	hdl = ds_pool_hdl_lookup(in->tci_hdl);
+	hdl = ds_pool_hdl_lookup(pic->pic_hdl);
 	if (hdl != NULL) {
-		if (hdl->sph_flags == in->tci_flags) {
+		if (hdl->sph_sec_capas == pic->pic_capas) {
 			D_DEBUG(DF_DSMS, DF_UUID": found compatible pool "
-				"handle: hdl="DF_UUID" flags="DF_U64"\n",
-				DP_UUID(in->tci_uuid), DP_UUID(in->tci_hdl),
-				hdl->sph_flags);
+				"handle: hdl="DF_UUID" capas="DF_U64"\n",
+				DP_UUID(pool->sp_uuid), DP_UUID(pic->pic_hdl),
+				hdl->sph_sec_capas);
 			rc = 0;
 		} else {
 			D_ERROR(DF_UUID": found conflicting pool handle: hdl="
-				DF_UUID" flags="DF_U64"\n",
-				DP_UUID(in->tci_uuid), DP_UUID(in->tci_hdl),
-				hdl->sph_flags);
+				DF_UUID" capas="DF_U64"\n",
+				DP_UUID(pool->sp_uuid), DP_UUID(pic->pic_hdl),
+				hdl->sph_sec_capas);
 			rc = -DER_EXIST;
 		}
 		ds_pool_hdl_put(hdl);
-		D_GOTO(out, rc);
+		return 0;
 	}
 
 	D_ALLOC_PTR(hdl);
 	if (hdl == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	pool = ds_pool_lookup(in->tci_uuid);
-	if (pool == NULL) {
-		D_FREE(hdl);
-		rc = -DER_NONEXIST;
-		goto out;
-	}
-
-	rc = ds_pool_tgt_map_update(pool, NULL, in->tci_map_version);
-	if (rc != 0) {
-		ds_pool_put(pool);
-		D_FREE(hdl);
-		goto out;
-	}
-
-	uuid_copy(hdl->sph_uuid, in->tci_hdl);
-	hdl->sph_flags = in->tci_flags;
-	hdl->sph_sec_capas = in->tci_sec_capas;
+	ds_pool_get(pool);
+	uuid_copy(hdl->sph_uuid, pic->pic_hdl);
+	hdl->sph_flags = pic->pic_flags;
+	hdl->sph_sec_capas = pic->pic_capas;
 	hdl->sph_pool = pool;
 
-	rc = daos_iov_copy(&hdl->sph_cred, &in->tci_cred);
-	if (rc != 0) {
-		ds_pool_put(pool);
-		D_FREE(hdl);
+	cred_iov.iov_len = pic->pic_cred_size;
+	cred_iov.iov_buf_len = pic->pic_cred_size;
+	cred_iov.iov_buf = &pic->pic_creds[0];
+	rc = daos_iov_copy(&hdl->sph_cred, &cred_iov);
+	if (rc != 0)
 		D_GOTO(out, rc);
-	}
 
 	rc = pool_hdl_add(hdl);
 	if (rc != 0) {
 		daos_iov_free(&hdl->sph_cred);
-		ds_pool_put(pool);
```

Comment:
ds_pool_put(pool) is still required, no?

Reply:
ah, yes.